### PR TITLE
Handle D-Bus Method Not Found When Systemd is Not the Current Init System

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

A bug was raised whereby:
- A series upgrade was attempted on a KVM container running Trusty.
- That system had a directory _/run/systemd/system_, which is used to detect systemd as the current init system.
- Calls to D-Bus method _LinkUnitFiles_ threw errors due to the method not being present.

This patch has 2 main commits.

The first simplifies the detection of systemd, using the upstream library's method instead of duplicating its logic.

The second has the logic change that handles the specific error and falls back to manually sym-linking service unit files. 

## QA steps

- Bootstrap and deploy a Trusty machine.
- SSH to the machine and create the directory _/run/systemd/system_.
- `juju upgrade-series 0 prepare Xenial`.
- Check the machine log and ensure that despite the systemd error messages, manually linking the agent service files follows and succeeds.

## Documentation changes

None.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1804701/

### CODE REVIEW

**Comment by manadart** [comment]
[Upgrade-Series Incorrect Usage of DBUS LinkUnitFiles](https://trello.com/c/Yb4ExXvW/827-upgrade-series-incorrect-usage-of-dbus-linkunitfiles)

File: service/agentconf.go [hunk 1]
```diff
@@ -192,23 +192,34 @@ func (s *systemdServiceManager) WriteSystemdAgents(
...
-			continue
+			// Note that this error is already logged by the systemd package.
+
+			// This is not ideal, but it is possible on an Upstart-based OS
+			// (such as Trusty) for run/systemd/system to exist, which is used
+			// for detection of systemd as the running init system.
+			// If this happens, then D-Bus will error with the message below.
+			// We need to detect this condition and fall through to linking the
+			// service files manually.
+			if strings.Contains(err.Error(), "No such method 'LinkUnitFiles'") {
```
Comment (jameinel) [inline service/agentconf.go]: At the very least, wouldn't we want to log the original error in the following Infof?
Also, do we know if this message is reasonably stable?
I assume there isn't an error code or any other clearer message than this that is *intentionally* giving the message, rather than accidentally.

I guess if we are only really focused on Trusty, whatever we have there isn't really going to be changing.

File: service/agentconf.go [hunk 2]
```diff
@@ -192,23 +192,34 @@ func (s *systemdServiceManager) WriteSystemdAgents(
...
-			return nil, nil, nil, errors.Errorf("failed to determine if systemd is running: %#v\n", err)
-		case running:
-			// If systemd is the running init system on this host, then the
-			// call to DBusAPI.LinkUnitFiles in WriteService above will have
-			// automatically updated sym-links for the file. We are done.
+		// If systemd is the running init system on this host, *and* if the
+		// call to DBusAPI.LinkUnitFiles in WriteService above returned no
+		// error, it will have resulted in updated sym-links for the file.
+		// We are done.
+		if s.isRunning() && !dbusMethodNotFound {
```
Comment (jameinel) [inline service/agentconf.go]: Using double negatives is a bit hard to follow here. Can we just use:
dbusMethodMissing
or 
dbusMethodFound  and invert the logic?

Comment (manadart) [reply service/agentconf.go]: Done.

File: service/agentconf.go (hunk 1 above)
Comment (manadart) [reply service/agentconf.go]: I agree; matching error strings smells. I looked through https://github.com/coreos/go-systemd and there is no typed error or code, for this.

Logged output looks like this:
```
machine-0: 10:37:31 ERROR juju.service.systemd dbus link request failed for application "jujud-machine-0": No such method 'LinkUnitFiles'
machine-0: 10:37:31 DEBUG juju.service.systemd stack trace:
No such method 'LinkUnitFiles'
/home/joseph/go/src/github.com/juju/juju/service/systemd/service.go:595: dbus link request failed for application "jujud-machine-0"
machine-0: 10:37:31 INFO juju.service attempting to manually link service file for machine-0
```

Comment (manadart) [reply service/agentconf.go]: I've made the matching a little softer, but I don't think there is much more we can do with it...

**Comment by manadart** [comment]
$$merge$$
